### ISSUE TITLE
Future development

### ISSUE THREAD

**Issue by saleem-hadad** [saleem-hadad/larecipe#6]
Here are some ideas for future improvement of this library:

1. RTL Support
2. Search Support
3. Notifications

Feel free to suggest yours

**Comment by viezel** [comment]
Regarding generating docs from inline docblocks: 
http://apidocjs.com is widely used. 
It generates a json output which can be converted to markdown. https://github.com/softdevstory/node-apidoc-markdown of course that needs to be adjusted to meet the requirements of this project. 

It would be a effective way to get adapters of Larecipe.

**Comment by johannesschobel** [comment]
Hey @saleem-hadad ,
wow, that is a cool project - i stumbled upon it by accident ;) Good work :+1: 

I wanted to make the same `DocBlock` suggestion - because my application already uses a lot of docblocks for internal documentation.. However, i would disagree with @viezel regarding `apidocJS` - the project is not active any more and some kind of discontinued as stated on the github page ( https://github.com/apidoc/apidoc ). Currently there are about 200 open issues and 25 PRs that are not merged (even small readme updates, ...) and open for about 2 years+ now.. 

For future perspective, at least from my current point of view - i would like to try `OpenAPI` (aka `Swagger`) as this seems very promising!

However, issue with `Swagger` is the huge docblock overload (e.g., docblock for every model attribute, ...) which is quite uncommon for laravel itself.. 

What are your opinions on this?

**Comment by julienbourdeau** [comment]
About Search support, I'd recommend Algolia 😄, and not only because I work there haha. Algolia powers search on the Laravel docs (as well as composer, vuejs, react, yarn, packagist, and more).

You can integrate via Laravel Scout, or you can get in touch with the DocSearch team to make it easy to integrate for your users. DocSearch is entirely free, as long as the doc is for an open source projet: https://community.algolia.com/docsearch/

Very nice project, congratulations! 🎉

**Comment by viezel** [comment]
Sure, Apidocjs is deprecated. My main suggestion was Docblock :) 

**API Formats**
* OpenAPI Spec 3.0 (JSON)
* OpenAPI/Swagger 2.0 JSON
* API Blueprint
* RAML 1.0

**Comment by johannesschobel** [comment]
i wouldn't use the `Swagger2` format, because it is already replaced by `OpenApi3` (Swagger 3) and the formats are incompatible! I once worked with swagger in a PHP project and the overall handling was quite aweful.. other languages, such as c# have a very (!) good built in swagger generator.. However, at the time looking at swagger2 and swagger3, the handling was not really that good.. .

I think, `API Blueprint` may be a valid alternative! I found this package quite nice for a hassle-free integration with laravel: https://github.com/M165437/laravel-blueprint-docs --> UI Generation

Cool thing with API Blueprint (apiary) is the tight integration with DREDD for automated API testing..

However, it looks like that you still need to "manually" write the `apib` file used by the blueprint-docs-generator package above.. Haven't found a "php code annotations to apib file" package ;)

**Comment by johannesschobel** [comment]
oh yeah.. another point in favor for `Api Blueprint`: Both (laracipe & apiblueprint) use markdown!

**Comment by saleem-hadad** [comment]
Thank you @johannesschobel for your feedback, my pleasure 👻
Well, to be honest, I'm not a big fan of Docblock to generate docs 🤣but I'll try my best in implementing or integrating other services that @viezel mentioned once I've free time since I'm working and studying at the same time 😵
Thank you guys again for your suggestions.

**Comment by saleem-hadad** [comment]
@julienbourdeau haha 😂, of course, Algolia is the best 🎊I was thinking of using Laravel Scout along with Algolia for sure.
I'll reach out DocSearch soon 👻
Thank you for your help and feedback.

**Comment by saleem-hadad** [comment]
Absolutely, If Docblocks implemented it gonna be as an optional plugin 👌

**Comment by ijpatricio** [comment]
hey all

@saleem-hadad 

take a look here [https://github.com/mpociot/documentarian](https://github.com/mpociot/documentarian)

Docblocks FTW 👍

**Comment by johannesschobel** [comment]
@ijpatricio I know this package.. It has some flaws you don't want to deal with.. further, it is also quite stale..

I would highly recommend to use an established standard, like API blueprint or others..

Cheers

**Comment by ijpatricio** [comment]
Thanks. I like the aglio.js one, but a Laravel and DocBlocks solution is a much greater fit for me.
I mean
- customizing something, publishing the views out from vendor and customize etc.
- doc blocks is just more natural to me, having it in one place

What kind of flaws? I just installed and played around a little, still to implement and use it for real...

**Comment by ijpatricio** [comment]
@saleem-hadad something i'm using now is vuepress, and wrap it in Laravel, serving the static asstes from a route and response with content and mimetype..

I would also gladly contribute, for snippet inclusion from file, as described here
[https://github.com/vuejs/vuepress/issues/750](https://github.com/vuejs/vuepress/issues/750)

**Comment by johannesschobel** [comment]
I dislike the way of the annotations.. Sure it works, I used the Laravel-doc-generator to generate the docs from code.. but didn't feel "natural" to me.. 

Further, I think (! Not sure !) that the format is "proprietary", i.e. It does not follow a standard.. 🤔

**Comment by ijpatricio** [comment]
Hi @saleem-hadad maybe you missed it, did you see the feature proposal about importing code snippets?

I see it very useful to technical docs, and usage without having to change the source code.

Something like
```
(```php)
echo 'LaRecipe is beatiful'; 
(```)
```
But with added feature:
- filePath: `relative to base path`, using later: `base_path('filePath')`
- first line query: `public function myMethod`
   (will find first line in file, with occurence of `public function myMethod`, and mark its line number)
- last line query: `return $something;`
   (will find first line in file, with occurence of `return $something;`, and mark its line number)
- lastLine > firstLine or throwException
- fileNumbering: `boolean`
   use the line numbers from file
- paddingTop/paddingBottom: `signedInteger`
   (padding to change first line and last linefor snippet) - useful for not having to place markers in source files
- highlightLinesQuery: [`some text`,`another text`]
   will highlight lines containing `some text` and `another text` 

I can check working this way working with prism, but also look below

I've done something similar with a vue component and CodeMirror, you can check [here](https://github.com/ijpatricio/nova-larastan-tool)

![codemirrorusage](https://github.com/ijpatricio/nova-larastan-tool/raw/master/docs/NovaLarastanTool.png)

What do you think?

**Comment by ijpatricio** [comment]
`paddingTop` and `paddingBottom` are different arguments, that default to 0 (zero).

**Comment by charlielangridge** [comment]
It would be great if we could 
1) Change the link back to the site on the nav-bar, (eg to /home instead of the site root)
2) If it was possible to override the views, so I can theme the docs to fit with the rest of the site

**Comment by baradhili** [comment]
It would be nice to have the same editor you have in the blogger package, so we can create docs in laravel, instead of having to push it up. 

Why? because in some cases, pushing code involves change windows/approvals etc.. editing docs would fall outside of that..

**Comment by saleem-hadad** [comment]
you mean providing a UI to edit them?
@baradhili

**Comment by saleem-hadad** [comment]
I like the idea

**Comment by rajmasha** [comment]
It would be great if we had navigation links to different project docs in the navbar itself. When a user clicks we navigate to /project1/docs/1.0/overview and the sidebar changes as per the project selected instead of having different sub-domains for different projects.

**Comment by saleem-hadad** [comment]
@rajmasha you mean multi-projects documentation?

**Comment by rajmasha** [comment]
Yeah exactly.

**Comment by kalynrobinson** [comment]
@saleem-hadad can you elaborate on what you had in mind for the feature `Notifications`? Alternatively, is there any feature you would like me to work on?

I also wanna 👍 docblocks support. I've seen similar front-end projects that include types and parameters via a React component, like `<ParamsTable for="Foo" />`, which I like as a simple compromise between those who docblock and those heretics who don't.

**Comment by saleem-hadad** [comment]
Thank you @kalynrobinson for your awesome help.
We use this package in our company and sometimes we need to notify our developers about changes in a project's specs or any requested modification, and so on. For that, I'm planning to have a small button in each page so that a developer can subscribe to get notifications if the page has been modified or new thing added.

**Comment by rbcunhadesign** [comment]
Is there a possibility of adding support for multiple docs?

For example, have one Laravel app with multiple documentation sets and each documentation set has its own directory/versions/search/links.

I am also using this for work (thanks for such an awesome package) but we are wanting to host all of our documentation under one auth system.

**Comment by rubensrocha** [comment]
> Is there a possibility of adding support for multiple docs?
> 
> For example, have one Laravel app with multiple documentation sets and each documentation set has its own directory/versions/search/links.
> 
> I am also using this for work (thanks for such an awesome package) but we are wanting to host all of our documentation under one auth system.

I also agree with this feature. Join all documentations in one LaRecipe installation is a good idea.

**Comment by abishekrsrikaanth** [comment]
I am looking forward to having a way to edit the docs through an editor instead of updating the source and pushing them everytime.

**Comment by saleem-hadad** [comment]
@abishekrsrikaanth  this feature will be available after supporting multi-languages

**Comment by rico** [comment]
I'm writing my API docs with larecipe - great software, thanks!

The main thing I'm missing is some kind of reusable data structures like in the [Blueprint](https://github.com/apiaryio/api-blueprint/blob/master/API%20Blueprint%20Specification.md#def-data-structures) specification.

I try to explain. I usually work with [API Resources](https://laravel.com/docs/5.7/eloquent-resources). Naturally, from time to time I have to add or change a field in these resources. Since I can use the resource in many responses  in my API (e.g. a user resource which is nested in another resource), I have to update the api docs in many places after such a change.

If I could, for instance, put my example responses for a resource in a dedicated file or block, and just reference it in my sample responses, it would make the process a lot easier ...

**Comment by jadamec** [comment]
+1 for Subdomain

**Comment by dekts** [comment]
when the editor will available for edit the docs.

**Comment by saleem-hadad** [comment]
@rico Thanks for sharing this and glad you enjoy working with Larecipe
Yeah, me too I'm facing the exact same issue, I use API resources a lot. in the next few days, I'll try to think in a nice solution for this issue

**Comment by AbbyJanke** [comment]
So right now if you have multiple themes installed, it will load the CSS Styling for all of them. Whichever one is loaded last takes precedence over the other ones. Example, using your dark theme I wanted to make some adjustments to it so created my own and only wanted to add styling for the changes i wanted but because it got loaded first (due to it going alphabetically) your dark one made my changes mute. I feel like a config for order they should be loaded would be awesome.

**Comment by saleem-hadad** [comment]
Nice Gotcha @AbbyJanke 
can you work on this?

**Comment by AbbyJanke** [comment]
@saleem-hadad PR #166 for the themes. Was either then I was expecting it to be in my head.

**Comment by robsontenorio** [comment]
UI editor would be nice, but it would require to implement some mechanism similar to a CMS, where all docs could be created from UI.

**Comment by WillGoldstein** [comment]
@robsontenorio like @saleem-hadad  has in blogged? I kinda think it’s not totally appropriate here because there’s so much more you can do here between vue components in markdown, the alerts boxes, etc, I feel like it would be just as easy as it is now, but probably really hard to build. Though a nice Ui might keep the docs easier to understand as you’re building them - like better structure layout.

**Comment by robsontenorio** [comment]
Basic idea: We can deal with creating `.md` files on server from UI, once LaRecipe is just a Laravel/PHP application. Right? So, it would be a powerfull combo!

**Comment by WillGoldstein** [comment]
Ah gotchya. Yeah gitbook is absurdly expensive. Like offensively expensive. Once I found this I was very happy with my choice to go with it. Getting it closer to what gitbook offers would be sweet.

**Comment by tsndp** [comment]
How about export full pledged pdf file from docs

**Comment by anthonyaxenov** [comment]
> Is there a possibility of adding support for multiple docs?
> 
> For example, have one Laravel app with multiple documentation sets and each documentation set has its own directory/versions/search/links.
> 
> I am also using this for work (thanks for such an awesome package) but we are wanting to host all of our documentation under one auth system.

Agree. It would be awesome to make something like this:
```php
<?php // src/config/larecipe.php
return [
    'docs'        => [
        [
            'route'   => '/docs/admin',
            'path'    => '/resources/docs/admin',
            'landing' => 'overview',
            'middleware' => ['web', 'auth'],
        ],[
            'route'   => '/docs/user',
            'path'    => '/resources/docs/user',
            'landing' => 'overview',
            'middleware' => ['web'],
        ],
    ]
    //...
];
```

**Comment by rfpdl** [comment]
> > Is there a possibility of adding support for multiple docs?
> > For example, have one Laravel app with multiple documentation sets and each documentation set has its own directory/versions/search/links.
> > I am also using this for work (thanks for such an awesome package) but we are wanting to host all of our documentation under one auth system.
> 
> Agree. It would be awesome to make something like this:
> 
> ```
> <?php // config/larecipe.php
> return [
>     'docs'        => [
>         [
>             'route'   => '/docs/admin',
>             'path'    => '/resources/docs/admin',
>             'landing' => 'overview',
>             'middleware' => ['web', 'auth'],
>         ],[
>             'route'   => '/docs/user',
>             'path'    => '/resources/docs/user',
>             'landing' => 'overview',
>             'middleware' => ['web'],
>         ],
>     ]
>     //...
> ];
> ```

This will definitely be a good feature. As we want to separate the documentation for end-users with internal users (who are managing the end users)

**Comment by saleem-hadad** [comment]
Hello guys, I hope you're doing well.

I've decided to come back to work on improving LaRecipe to the next level. However, I'm having difficulty balancing my new married life, a full-time job, and open-source development.
Therefore, I'm thinking if I want to come back working on LaRecipe and adding more features I need to be more motivated since I'll invest my spear time on the product. 

So, to be honest, I'm planning to release a pro version of LaRecipe with a one-lifetime payment for unlimited projects and that will include 6 months of professional support and unlimited major version updates, and here I just want to know if you're interested in the new pro version you can register your interest [here](https://larecipe.binarytorch.com.my/docs/pro/overview) to get notified.

Now, I know this might be something you don't like but the good news is that the free version will continue to be maintained for adding small/minor features and bug fixes.

Best regards.

**Comment by WillGoldstein** [comment]
@saleem-hadad Heck yes. I’d pay. Keep this thing going. Love using it.

**Comment by saleem-hadad** [comment]
I opened a new chat server, please do join if you're interested in the project's updates:
https://discord.gg/DktCdh

**Comment by e-hop** [comment]
i wouldn't discard this package if it was using react. It'll be great to have a react version.

**Comment by danFWD** [comment]
Hello, great package! Has there been any progress made on support for multiple docs? I see that this was added to the To Do list, but I don’t see any documentation for it, and when I tried to use multiple configuration arrays for the `docs` routing config, it threw an error. Thank you for your excellent work!
